File: src/edu/uf/interactable/TNFa.java
package edu.uf.interactable;

import edu.uf.Diffusion.Diffuse;
import edu.uf.compartments.GridFactory;
import edu.uf.intracellularState.Phenotype;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants;
import edu.uf.utils.Util;

public class TNFa extends Molecule{

	public static final String NAME = "TNFa";
	public static final int NUM_STATES = 1;
	
	private static TNFa molecule = null;
    
    private TNFa(double[][][][] qttys, Diffuse diffuse) {
		super(qttys, diffuse);
		this.setPhenotye(Phenotype.createPhenotype());
	}
    
    public static TNFa getMolecule(Diffuse diffuse) {
    	if(molecule == null) {
    		double[][][][] values = new double[NUM_STATES][GridFactory.getXbin()][GridFactory.getYbin()][GridFactory.getZbin()];
    		molecule = new TNFa(values, diffuse); 
    	}
    	return molecule;
    }
    
    public static TNFa getMolecule() {
    	return getMolecule(null);
    }
    
    @Override
    public double getKd() {
    	return Constants.Kd_TNF;
    }
    
    public void degrade() {
    	degrade(Constants.TNF_HALF_LIFE, 0);
    }

    public int getIndex(String str) {
        return 0;
    }

    public void computeTotalMolecule(int x, int y, int z) {
    	this.totalMoleculesAux[0] = this.totalMoleculesAux[0] + this.get(0, x, y, z);
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof Macrophage) {
        	Interactions.secrete((Leukocyte) interactable, this, Constants.MA_TNF_QTTY, x, y, z, 0);
            return Interactions.bind((Leukocyte) interactable, this, x, y, z, 0);
        }
        if (interactable instanceof Neutrophil) { 
        	Interactions.secrete((Leukocyte) interactable, this, Constants.N_TNF_QTTY, x, y, z, 0);
        	return Interactions.bind((Leukocyte) interactable, this, x, y, z, 0);
        }
        return interactable.interact(this, x, y, z);
    }

	@Override
	public String getName() {
		return NAME;
	}

	@Override
	public double getThreshold() {
		return 0;
	}

	@Override
	public int getNumState() {
		return NUM_STATES;
	}
	
	@Override
	public boolean isTime() {
		return true;
	}
}

File: cpp/src/edu/uf/interactable/TNFa.cpp
/*
 * TNFa.cpp
 *
 *  Created on: Jul 17, 2024
 *      Author: henriquedeassis
 */

#include "TNFa.h"
#include "../utils/Id.h"
#include "../utils/Constants.h"
#include "../utils/Util.h"
#include "Macrophage.h"
#include <memory>

namespace edu {
namespace uf {
namespace interactable {

using namespace edu::uf::utils;
//using namespace primitives;

const int TNFa::classId = Id::getId();
const string TNFa::NAME = "TNFa";
const int TNFa::NUM_STATES = 1;
TNFa* TNFa::molecule = nullptr;

TNFa::TNFa(double**** qttys, int xbin, int ybin, int zbin, char wbin, Diffuse* diffuse)
    : Molecule(qttys, xbin, ybin, zbin, wbin, diffuse) {
    setPhenotype(Id::getId());
}

TNFa* TNFa::getMolecule(Diffuse* diffuse, int xbin, int ybin, int zbin, char wbin) {
    if (molecule == nullptr) {
        double**** values = new double***[wbin];
        for (int i = 0; i < wbin; ++i) {
            values[i] = new double**[xbin];
            for (int j = 0; j < xbin; ++j) {
                values[i][j] = new double*[ybin];
                for (int k = 0; k < ybin; ++k) {
                    values[i][j][k] = new double[zbin]();
                }
            }
        }
        molecule = new TNFa(values, xbin, ybin, zbin, wbin, diffuse);
    }
    return molecule;
}

int TNFa::getClassId() const {
	return classId;
}

double TNFa::getKd() {
    return utils::constexprants::Kd_TNF;
}

void TNFa::degrade() {
    //degrade(utils::constexprants::TNF_HALF_LIFE, 0);
}

int TNFa::getIndex(const std::string& str) const {
    return 0;
}

void TNFa::computeTotalMolecule(int x, int y, int z) {
    totalMoleculesAux[0] += get(0, x, y, z);
}

bool TNFa::templateInteract(Interactable* interactable, int x, int y, int z) {
    if (dynamic_cast<Macrophage*>(interactable)) {
    	Cell* cell = static_cast<Cell*>(interactable);
    	if(cell->getBooleanNetwork()->hasPhenotype(*this)){
    		double level = cell->getBooleanNetwork()->getPhenotype().at(this->getPhenotype())/4.0;
    		this->inc(utils::constexprants::MA_TNF_QTTY * level, 0, x, y, z);
    	}
    	cell->bind(molecule, utils::Util::activationFunction5(this->get(0, x, y, z), this->getKd()));
        return true;
    }
    return interactable->interact(this, x, y, z);
}

string TNFa::getName() const {
    return NAME;
}

double TNFa::getThreshold() const {
    return 0;
}

int TNFa::getNumState() const {
    return NUM_STATES;
}

bool TNFa::isTime() const {
    return true;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */
